### DISCUSSION TITLE
UART usage on RP2040: UART.IRQ, run UART on CORE_1

### DISCUSSION THREAD

**Discussion by Flipje1955** [micropython/micropython#12247] [RP2040 / Pico]
As part of my project, I need to process a load of external sensor to detect passing objects (mini_train project). The sensors are typically but not exclusively hall_sensor modules. I have dedicated a rp2040 Pico to control all these sensors in order to unload the "master controller", also running on a rp2040 Pico. 
I intend to have the two (and in the future ++) rp2040's talk to each other using UARTs where I had hoped to use UART.irq to avoid continued polling for incoming data but that functionality appears not to be implemented (yet ?) for this platform. 
I have seen recommendations to use Asyncio instead.

Considering my options (and maintaining the ambition to unload the master), is using the Threading  library and "dedicating" the Core1 to the UART monitoring for any incoming sensor_data a possible and sensible route ? 
Thank you.

**Comment by peterhinch** [comment]
I suggest ignoring core 1 and using `asyncio` throughout. If you look at [the tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md) you will see that UARTs are handled easily using `StreamReader` and `StreamWriter` objects.

Multi-core programming is difficult with a lot of potential for subtle bugs. Using `asyncio` is much easier despite the learning curve.

**Reply by Flipje1955** [reply]
thx, Peter. As part of my project I have used aysncio to produce a 4 channel sound generator to simulate train sounds, building on Mike Teachman's work, using I2S / asyncio and benefitted from your tutorial where I noted the UART example. Wrt asyncio, I created a PCB with two rp2040's pico, each with an SDcard, allowing me to play a variety of WAV_files.  Interfacing is over UART. So that experience I can put to work. On threading, the sensor module I created is also based on a pico and where Core1 is used to show sensor data on the locally attached led display. Did not encounter any major problems but I do believe that in (more) complex use cases, problems will surface. 
For my project, running multiple trains in parallel, I am sure I must switch to asyncio at some point but hope(d) to delay that until I have solved a series of other challenges, incl. further improving my design and (MPY) programming skills. Again, thx for advice !

**Comment by shariltumin** [comment]
@Flipje1955, hey. You might find this interesting:

```
MicroPython v1.20.0 on 2023-04-26; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> from machine import UART, Pin
>>> uart = UART(1, baudrate=9600, tx=Pin(4), rx=Pin(5))
>>> uart.deinit()
>>> uart.init()
>>> def IRQ(pin):
...     w = uart.read()
...     if w:
...        print(w)
... 
>>> Pin(5).irq(IRQ, trigger=Pin.IRQ_FALLING)
<irq>
>>> uart.write('hello')
5
>>> b'hello'

>>> uart.write('bye')
3
>>> b'bye'

>>>
```

To test this, you need to short GPIO4 and GPIO5 (on the same board) of your Pi Pico.

You may get some noise and need to filter it out in your handler.
You might want to put ```uart.deinit();uart.init()``` after powerup.

I do not know how useful or stable this setup is for your project.
It is safe if all your boards have GPIO with 3V3 output.

You can give it a try. There is always *asyncio* to fall back on to.

**Reply by Flipje1955** [reply]
Thx for your suggestion ! Will give it a try but it should work (signal levels are all 3.3V); practical challenge is that I am already short on GPIO pins.....
